modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxRecoveryConcurrentOnPrimaryFailTest.java

```diff
@@ -0,0 +1,171 @@
+    }
+
+    /**
+     * The test enforces the concurrent processing of the same prepared transaction
+     * both in the tx recovery procedure started due to primary node left and in the
+     * tx recovery request handler invoked by message from another backup node.
+     * <p>
+     * The idea is to have a 3-nodes cluster and a cache with 2 backups. So there
+     * will be 2 backup nodes to execute the tx recovery in parallel if primary one
+     * would fail. These backup nodes will send the tx recovery requests to each
+     * other, so the tx recovery request handler will be invoked as well.
+     * <p>
+     * Use several attempts to reproduce the race condition.
+     * <p>
+     * Expected result: transaction is finished on both backup nodes and the partition
+     * map exchange is completed as well.
+     */
+    @Test
+    public void testRecoveryNotDeadLockOnPrimaryFail() throws Exception {
+        final IgniteEx grid0 = startGrid(0);
+
+        final IgniteEx grid1 = startGrid(1);
+
+        final CyclicBarrier grid1BlockerBarrier = new CyclicBarrier(3);
+
+        final Runnable grid1BlockerTask = () -> {
+            try {
+                grid1BlockerBarrier.await();
+            }
+            catch (InterruptedException | BrokenBarrierException e) {
+                // Just supress.
+            }
+        };
+
+        for (int iter = 0; iter < 100; iter++) {
+            final IgniteEx grid2 = startGrid(2);
+
+            awaitPartitionMapExchange();
+
+            final IgniteCache<Object, Object> cache = grid2.cache(DEFAULT_CACHE_NAME);
+
+            final Transaction tx = grid2.transactions().txStart(PESSIMISTIC, REPEATABLE_READ);
+
+            // Key for which the grid2 node is primary.
+            final Integer grid2PrimaryKey = primaryKeys(cache, 1, 0).get(0);
+
+            cache.put(grid2PrimaryKey, Boolean.TRUE);
+
+            final TransactionProxyImpl<?, ?> p = (TransactionProxyImpl<?, ?>)tx;
+
+            p.tx().prepare(true);
+
+            final Collection<IgniteInternalTx> txs0 = grid0.context().cache().context().tm().activeTransactions();
+            final Collection<IgniteInternalTx> txs1 = grid1.context().cache().context().tm().activeTransactions();
+            final Collection<IgniteInternalTx> txs2 = grid2.context().cache().context().tm().activeTransactions();
+
+            assertTrue(txs0.size() == 1);
+            assertTrue(txs1.size() == 1);
+            assertTrue(txs2.size() == 1);
+
+            final IgniteInternalTx tx0 = txs0.iterator().next();
+            final IgniteInternalTx tx1 = txs1.iterator().next();
+
+            // Block recovery procedure processing on grid1.
+            grid1.context().pools().getSystemExecutorService().execute(grid1BlockerTask);
+
+            // Block stripe tx recovery request processing on grid1 (note that the only stripe is configured in the executor).
+            grid1.context().pools().getStripedExecutorService().execute(0, grid1BlockerTask);
```

Comment:
Please make clear what role has the node where you're blocking pools

Reply:
Fixed